pageserver/src/tenant/ephemeral_file.rs

```diff
@@ -198,50 +198,54 @@ impl FileExt for EphemeralFile {
 
 impl BlobWriter for EphemeralFile {
     fn write_blob(&mut self, srcbuf: &[u8]) -> Result<u64, io::Error> {
-        let pos = self.size;
-
-        let mut blknum = (self.size / PAGE_SZ as u64) as u32;
-        let mut off = (pos % PAGE_SZ as u64) as usize;
+        struct Writer<'a> {
+            ephemeral_file: &'a mut EphemeralFile,
+            blknum: u32,
+            off: usize,
+        }
+        impl<'a> Writer<'a> {
+            fn new(ephemeral_file: &'a mut EphemeralFile) -> Writer<'a> {
+                Writer {
+                    blknum: (ephemeral_file.size / PAGE_SZ as u64) as u32,
+                    off: (ephemeral_file.size % PAGE_SZ as u64) as usize,
+                    ephemeral_file,
+                }
+            }
+            fn push_bytes(&mut self, src: &[u8]) -> Result<(), io::Error> {
+                let mut src_remaining = src;
+                while !src_remaining.is_empty() {
+                    {
+                        let mut head_page = self.ephemeral_file.get_buf_for_write(self.blknum)?;
+                        let dst_remaining = &mut head_page[self.off..];
+                        let n = min(dst_remaining.len(), src_remaining.len());
+                        dst_remaining[..n].copy_from_slice(&src_remaining[..n]);
+                        self.off += n;
+                        src_remaining = &src_remaining[n..];
+                    }
+                    if self.off == PAGE_SZ {
+                        // This block is done, move to next one.
+                        self.blknum += 1;
+                        self.off = 0;
+                    }
+                }
+                Ok(())
+            }
+        }
 
-        let mut buf = self.get_buf_for_write(blknum)?;
+        let pos = self.size;
+        let mut writer = Writer::new(self);
 
         // Write the length field
         if srcbuf.len() < 0x80 {
-            buf[off] = srcbuf.len() as u8;
-            off += 1;
+            writer.push_bytes(&[srcbuf.len() as u8])?;
```

Comment:
My brain has trouble parsing this. Maybe:

```suggestion
            // short one-byte length header
            let len_buf: &u8[] = &[srcbuf.len() as u8];
            writer.push_bytes(&len_buf)?;
```

Reply:
4beac4d4249015e4e81a6ea3710f1affda35a40b